blockchain/ethereum.go

```diff
@@ -536,6 +537,69 @@ func (e *EthereumClient) ProcessEvent(name string, event *types.Log, confirmedCh
 	return nil
 }
 
+// PollFinalizedHeader continuously polls the latest finalized header and stores it in the client
+func (e *EthereumClient) PollFinality() {
+	e.FinalizedHeader.Store(newGlobalFinalizedHeaderManager(e))
+}
+
+// StopPollingForFinalizedHeader stops polling for the latest finalized header
+func (e *EthereumClient) CancelFinalityPolling() {
+	if _, ok := e.headerSubscriptions[finalizedHeaderKey]; ok {
+		e.DeleteHeaderEventSubscription(finalizedHeaderKey)
+	}
+}
+
+// WaitForTxTobeFinalized waits for a transaction to be finalized
+// If the network is simulated, it will return immediately
+// otherwise it waits for the transaction to be finalized and returns the block number and time of the finalization
+func (e *EthereumClient) WaitForFinalizedTx(txHash common.Hash) (*big.Int, time.Time, error) {
+	if e.NetworkSimulated() {
+		return nil, time.Time{}, nil
+	}
+
+	receipt, err := e.Client.TransactionReceipt(context.Background(), txHash)
+	if err != nil {
+		return nil, time.Time{}, err
+	}
+	txHdr, err := e.HeaderByNumber(context.Background(), receipt.BlockNumber)
+	if err != nil {
+		return nil, time.Time{}, err
+	}
+	finalizer := NewTransactionFinalizer(e, txHdr, receipt.TxHash)
+	key := "txFinalizer-" + txHash.String()
+	e.AddHeaderEventSubscription(key, finalizer)
+	defer e.DeleteHeaderEventSubscription(key)
+	err = finalizer.Wait()
```

Comment:
Are there situations where you want to wait for multiple finalized txs and then call the general `WaitForEvents`?

Reply:
Including multiple transactions for finality check can lead to unpredictable delays. That's why not including it in `WaitForEvents` was a conscious decision. It's clean to do it by individual transactions. You generally wouldn't want to to wait for finality for generic transactions.